File: backend/app/schemas/generations.py
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class GenerationCostumeResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    name: str
    parent_id: str | None = Field(default=None, alias="parentId")
    costume_lora: str | None = Field(default=None, alias="costumeLora")
    costume_prompt: str = Field(alias="costumePrompt")
    created_at: str = Field(alias="createdAt")
    preview_images: list[str] = Field(default_factory=list, alias="previewImages")


class GenerationPromptSources(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    dna_prompt: str = Field(alias="dnaPrompt")
    trigger_word: str = Field(alias="triggerWord")
    costume_prompt: str = Field(alias="costumePrompt")


class GenerationParameterDefaults(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    width: int
    height: int
    steps: int
    sampler: str
    cfg_scale: float = Field(alias="cfgScale")
    seed: int
    lora_weight: float = Field(alias="loraWeight")


class GenerationWorkbenchResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    character_id: str = Field(alias="characterId")
    character_name: str = Field(alias="characterName")
    can_generate: bool = Field(alias="canGenerate")
    blocking_reason: str | None = Field(default=None, alias="blockingReason")
    costumes: list[GenerationCostumeResponse]
    selected_costume_id: str = Field(alias="selectedCostumeId")
    prompt_sources: GenerationPromptSources = Field(alias="promptSources")
    parameter_defaults: GenerationParameterDefaults = Field(alias="parameterDefaults")
    tag_options: list[str] = Field(
        default_factory=lambda: ["封面图", "表情包", "周边", "预告图"],
        alias="tagOptions",
    )


class GenerationSubmitRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    character_id: str = Field(alias="characterId")
    costume_id: str = Field(alias="costumeId")
    scene_prompt: str = Field(min_length=1, alias="scenePrompt")
    negative_prompt: str = Field(default="", alias="negativePrompt")
    width: int = Field(ge=64, le=4096)
    height: int = Field(ge=64, le=4096)
    steps: int = Field(ge=1, le=150)
    sampler: str = Field(min_length=1)
    cfg_scale: float = Field(ge=0.0, le=30.0, alias="cfgScale")
    seed: int | None = None
    lora_weight: float = Field(ge=0.0, le=2.0, alias="loraWeight")
    tags: list[str] = Field(default_factory=list)


class GenerationMockJobResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    task_id: str = Field(alias="taskId")
    character_id: str = Field(alias="characterId")
    costume_id: str = Field(alias="costumeId")
    scene_prompt: str = Field(alias="scenePrompt")
    status: Literal["pending", "running", "completed", "failed"]
    stage: Literal["queued", "running", "completed", "failed"]
    progress: int = Field(ge=0, le=100)
    message: str | None = None
    error: str | None = None
    tags: list[str] = Field(default_factory=list)
    created_at: str = Field(alias="createdAt")
    updated_at: str = Field(alias="updatedAt")


class GenerationJobAcceptedResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    job: GenerationMockJobResponse

File: src/api/generations.ts
import type { TaskSnapshot } from "./tasks";

export type GenerationCostumeOption = {
  id: string;
  name: string;
  costumePrompt: string;
  isDefault: boolean;
};

export type GenerationPromptSources = {
  dnaPrompt: string;
  triggerWord: string;
  costumePrompt: string;
};

export type GenerationParameterDefaults = {
  width: number;
  height: number;
  steps: number;
  sampler: string;
  cfgScale: number;
  seed: number | null;
  loraWeight: number;
};

export type GenerationWorkbenchContract = {
  characterId: string;
  characterName: string;
  canGenerate: boolean;
  blockingReason: string | null;
  costumes: GenerationCostumeOption[];
  selectedCostumeId: string;
  promptSources: GenerationPromptSources;
  parameterDefaults: GenerationParameterDefaults;
  tagOptions: string[];
};

export type GenerationMockRequest = {
  characterId: string;
  costumeId: string;
  scenePrompt: string;
  negativePrompt: string;
  width: number;
  height: number;
  steps: number;
  sampler: string;
  cfgScale: number;
  seed: number | null;
  loraWeight: number;
  tags: string[];
};

export type GenerationMockJob = {
  id: string;
  taskId: string;
  characterId: string;
  costumeId: string;
  scenePrompt: string;
  status: "pending" | "running" | "completed" | "failed";
  stage: "queued" | "running" | "completed" | "failed";
  progress: number;
  message?: string | null;
  error?: string | null;
  tags: string[];
  createdAt: string;
  updatedAt: string;
};

type GenerationCostumePayload = {
  id?: unknown;
  name?: unknown;
  costumePrompt?: unknown;
  isDefault?: unknown;
  parentId?: unknown;
};

const DEFAULT_API_BASE_URL = "http://127.0.0.1:8000";

function resolveWorkbenchUrl(characterId: string): string {
  return `${DEFAULT_API_BASE_URL}/api/characters/${characterId}/generation-workbench`;
}

function resolveMockGenerationUrl(): string {
  return `${DEFAULT_API_BASE_URL}/api/generations/mock`;
}

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

function normalizeGenerationCostume(payload: unknown): GenerationCostumeOption | null {
  if (!isObject(payload)) {
    return null;
  }

  const candidate = payload as GenerationCostumePayload;
  if (
    typeof candidate.id !== "string" ||
    typeof candidate.name !== "string" ||
    typeof candidate.costumePrompt !== "string"
  ) {
    return null;
  }

  const isDefault =
    typeof candidate.isDefault === "boolean"
      ? candidate.isDefault
      : candidate.parentId === null && candidate.name === "基础造型";

  return {
    id: candidate.id,
    name: candidate.name,
    costumePrompt: candidate.costumePrompt,
    isDefault,
  };
}

function normalizeGenerationWorkbenchContract(value: unknown): GenerationWorkbenchContract | null {
  if (!isObject(value)) {
    return null;
  }

  const costumes = Array.isArray(value.costumes)
    ? value.costumes.map((item) => normalizeGenerationCostume(item))
    : null;
  if (costumes === null || costumes.some((item) => item === null)) {
    return null;
  }

  if (
    typeof value.characterId !== "string" ||
    typeof value.characterName !== "string" ||
    typeof value.canGenerate !== "boolean" ||
    (value.blockingReason !== null && typeof value.blockingReason !== "string") ||
    typeof value.selectedCostumeId !== "string" ||
    !isObject(value.promptSources) ||
    !isObject(value.parameterDefaults) ||
    !Array.isArray(value.tagOptions) ||
    !value.tagOptions.every((item) => typeof item === "string")
  ) {
    return null;
  }

  const promptSources = value.promptSources;
  const parameterDefaults = value.parameterDefaults;
  if (
    typeof promptSources.dnaPrompt !== "string" ||
    typeof promptSources.triggerWord !== "string" ||
    typeof promptSources.costumePrompt !== "string" ||
    typeof parameterDefaults.width !== "number" ||
    typeof parameterDefaults.height !== "number" ||
    typeof parameterDefaults.steps !== "number" ||
    typeof parameterDefaults.sampler !== "string" ||
    typeof parameterDefaults.cfgScale !== "number" ||
    (parameterDefaults.seed !== null && typeof parameterDefaults.seed !== "number") ||
    typeof parameterDefaults.loraWeight !== "number"
  ) {
    return null;
  }

  return {
    characterId: value.characterId,
    characterName: value.characterName,
    canGenerate: value.canGenerate,
    blockingReason: value.blockingReason,
    costumes: costumes as GenerationCostumeOption[],
    selectedCostumeId: value.selectedCostumeId,
    promptSources: {
      dnaPrompt: promptSources.dnaPrompt,
      triggerWord: promptSources.triggerWord,
      costumePrompt: promptSources.costumePrompt,
    },
    parameterDefaults: {
      width: parameterDefaults.width,
      height: parameterDefaults.height,
      steps: parameterDefaults.steps,
      sampler: parameterDefaults.sampler,
      cfgScale: parameterDefaults.cfgScale,
      seed: parameterDefaults.seed as number | null,
      loraWeight: parameterDefaults.loraWeight,
    },
    tagOptions: value.tagOptions as string[],
  };
}

function isGenerationMockJob(value: unknown): value is GenerationMockJob {
  if (!isObject(value)) {
    return false;
  }

  const validStatus =
    value.status === "pending" ||
    value.status === "running" ||
    value.status === "completed" ||
    value.status === "failed";
  const validStage =
    value.stage === "queued" ||
    value.stage === "running" ||
    value.stage === "completed" ||
    value.stage === "failed";

  return (
    typeof value.id === "string" &&
    typeof value.taskId === "string" &&
    typeof value.characterId === "string" &&
    typeof value.costumeId === "string" &&
    typeof value.scenePrompt === "string" &&
    validStatus &&
    validStage &&
    typeof value.progress === "number" &&
    Array.isArray(value.tags) &&
    value.tags.every((item) => typeof item === "string") &&
    typeof value.createdAt === "string" &&
    typeof value.updatedAt === "string"
  );
}

function isGenerationMockResponse(value: unknown): value is { job: GenerationMockJob } {
  if (!isObject(value)) {
    return false;
  }

  return isGenerationMockJob(value.job);
}

async function readJson(response: Response): Promise<unknown> {
  try {
    return await response.json();
  } catch {
    return null;
  }
}

function readBackendDetail(payload: unknown): string | null {
  if (!isObject(payload)) {
    return null;
  }

  return typeof payload.detail === "string" ? payload.detail : null;
}

export async function fetchGenerationWorkbenchContract(
  characterId: string,
  signal?: AbortSignal,
): Promise<GenerationWorkbenchContract> {
  let response: Response;
  try {
    response = await fetch(resolveWorkbenchUrl(characterId), { signal });
  } catch {
    throw new Error("GENERATION_CONTRACT_UNAVAILABLE");
  }

  const payload = await readJson(response);

  if (!response.ok) {
    throw new Error(readBackendDetail(payload) ?? "GENERATION_CONTRACT_UNAVAILABLE");
  }

  const contract = normalizeGenerationWorkbenchContract(payload);
  if (contract === null) {
    throw new Error("INVALID_GENERATION_CONTRACT_RESPONSE");
  }

  return contract;
}

export async function createMockGenerationJob(
  input: GenerationMockRequest,
  signal?: AbortSignal,
): Promise<GenerationMockJob> {
  let response: Response;
  try {
    response = await fetch(resolveMockGenerationUrl(), {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(input),
      signal,
    });
  } catch {
    throw new Error("GENERATION_SUBMIT_UNAVAILABLE");
  }

  const payload = await readJson(response);

  if (!response.ok) {
    throw new Error(readBackendDetail(payload) ?? "GENERATION_SUBMIT_FAILED");
  }

  if (!isGenerationMockResponse(payload)) {
    throw new Error("INVALID_GENERATION_SUBMIT_RESPONSE");
  }

  return payload.job;
}

export function mergeTaskIntoGenerationJob(
  job: GenerationMockJob,
  task: TaskSnapshot,
): GenerationMockJob {
  const stage =
    task.status === "pending"
      ? "queued"
      : task.status === "running"
        ? "running"
        : task.status === "completed"
          ? "completed"
          : "failed";

  return {
    ...job,
    status: task.status,
    stage,
    progress: task.progress,
    message: task.message,
    error: task.error,
    updatedAt: task.updatedAt,
  };
}
